### PR TITLE
Replace CGLIB with manual wrappers which notify event listeners

### LINKED ISSUE #327: Getting rid of CGLIB
In this issue I want to discuss the pros and cons to remove CGLIB and to base p6spy on manual wrapper objects for the main `java.sql` interfaces.

Pros:
* Performance
 * CGLIB was the cause of a major performance problem (see #325). Although the latest version of CGLIB seems to handle concurrency better there can still be contention as the synchronized blocks are not removed completely.
 * Right now, every time a JDBC call is about to happen, multiple classes have to be created - even if the proxy has already been generated for the underlying class.
 * By using manual wrappers there is no runtime overhead for creating proxies and multithreading does not harm the performance
* Plays nicely with the event architecture discussed in #319
* Smaller jar. The majority of the p6spy jar consists of the shaded cglib dependency. Without it, p6spy would be more lightweight.
* Improved maintainability
 * Problems related to byte code manipulation are usually harder to debug and to resolve.
 * Less magical code. I'd consider CGLIB/byte code manipulation as a advanced topic. Thus it's harder for new commiters to understand the codebase. The new approach would be 100% Java SE
 * No danger of `ClassNotFoundException`s because of class loader issues

Cons:
* Casting the `Connection` to a vendor specific class would fail. But [unwrap](http://docs.oracle.com/javase/7/docs/api/java/sql/Wrapper.html#unwrap(java.lang.Class))ing the connection would still work and is preferred anyway.
* If the `java.sql` interfaces are extended, p6spy would have to support the new methods in order to work with the latest Java versions. Adding the new methods wouldn't be a lot of hassle though.

I'm currently working on this and would appreciate feedback:

https://github.com/felixbarny/p6spy/commits/event-emitting-wrapper

### PR DESCRIPTION
**Pull request by felixbarny**
Closes #327 

No regressions in integration tests: https://travis-ci.org/felixbarny/p6spy-it/builds/139815516

### CODE REVIEW

**Comment by felixbarny** [comment]
In fact, web logic integration tests are working again!

**Comment by felixbarny** [comment]
Also closes #319, #298
Support for #299 is easy to add

**Comment by quintonm** [comment]
@felixbarny Do you have changes for the p6spy-it repo?  The WLS build is failing in our repo due to a setup error.  I can't see how code changes to the p6spy project would fix that.

File: src/main/java/com/p6spy/engine/wrapper/ConnectionWrapper.java [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+  public Array createArrayOf(String typeName, Object[] elements) throws SQLException {
+    return delegate.createArrayOf(typeName, elements);
+  }
+
+  @Override
+  public Struct createStruct(String typeName, Object[] attributes) throws SQLException {
+    return delegate.createStruct(typeName, attributes);
+  }
+
+  public void setSchema(String schema) throws SQLException {
```
Comment (quintonm) [inline src/main/java/com/p6spy/engine/wrapper/ConnectionWrapper.java]: Minor - missing a few @Override annotations here

**Comment by quintonm** [comment]
This looks really good.  I am still a bit concerned about the loss of access to vendor specific methods though.  Maybe we just wait and see if it becomes an issue.   

BTW - There are a few conflicts to resolve before it can be merged.

**Comment by felixbarny** [comment]
The only thing that I have changed in the integration test setup is that it checks out my fork and installs the maven project:

```
before_install:
  - jdk_switcher use oraclejdk8
  - git clone https://github.com/felixbarny/p6spy.git
  - pushd p6spy
  - git checkout event-emitting-wrapper
  - mvn clean install -DskipTests=true
  - popd
  - jdk_switcher use oraclejdk7
  - export ORG_GRADLE_PROJECT_container=${AS}
```

> I am still a bit concerned about the loss of access to vendor specific methods though

To be honest, I'm a bit concerned as well. But most of the times this is a client error I think and can be solved by unwrapping the connection. If the error is inside an open source project (like an application server or liquibase), we can always create a pull request in those projects. But we can always choose to reintroduce CGLIB or to use Byte Buddy. Maybe that could also be in a different maven module so that when having a dependency on `p6spy-classgen` it does not use wrapper classes but generated classes.

@typekpb could you compare the performance between the master and my pull requests?

I'll add the `@Override` annotations and implement #299. How do you feel about javadoc in the wrapper classes? I don't feel that method level javadoc adds any value..

**Comment by typekpb** [comment]
@felixbarny well, regarding perf test I'm superbusy these days. So start of
july would be the earliest real date for me.

**Comment by felixbarny** [comment]
The performance tests where very easy to execute. Good work on that @typekpb! 

I can definitely see improvements:

2.3.1:

uri | count | total | min | average | max | standarddeviation | persecond | errors
--- | ----- | ----- | --- | ------- | --- | ----------------- | --------- | ------
"no-p6spy select" | 10 | 2121 | 154 | 212 | 320 | 61 | 1 | 0
"p6spy select" | 10 | 6760 | 522 | 676 | 898 | 109 | 1 | 0

Wrappers:

uri | count | total | min | average | max | standarddeviation | persecond | errors
--- | ----- | ----- | --- | ------- | --- | ----------------- | --------- | ------
"no-p6spy select" | 10 | 2070 | 152 | 207 | 328 | 63 | 1 | 0
"p6spy select" | 10 | 4168 | 332 | 416 | 528 | 69 | 2 | 0

**Comment by felixbarny** [comment]
The event architecture will also make the stagemonitor integration much more efficient. Most of the overhead in the logging module comes from the onResultSetGet and onResultSetNext callbacks. As stagemonitor is not interested in that information, there is almost no overhead when calling an empty event listener method - especially after the JIT has kicked in.

**Comment by felixbarny** [comment]
Anyone dare to merge?

**Comment by quintonm** [comment]
LGTM

**Pull request merged**